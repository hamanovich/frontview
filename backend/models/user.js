import mongoose from 'mongoose';
import md5 from 'md5';
import jwt from 'jsonwebtoken';
import uniqueValidator from 'mongoose-unique-validator';

const { Schema } = mongoose;

const userSchema = new Schema(
  {
    username: {
      type: String,
      unique: true,
      trim: true,
      required: 'Username field is required',
    },
    email: {
      type: String,
      unique: true,
      lowercase: true,
      trim: true,
      required: 'Email field is required',
    },
    passwordDigest: {
      type: String,
      required: 'Password field is required',
    },
    firstName: String,
    lastName: String,
    primarySkill: String,
    jobFunction: String,
    skype: String,
    phone: String,
    notes: {
      type: String,
      default: 'Please add some notes about yourself',
    },
    role: {
      type: String,
      enum: ['user', 'owner', 'admin'],
      default: 'user',
    },
    resetPasswordToken: String,
    resetPasswordExpires: Date,
    confirmed: {
      type: Boolean,
      default: false,
    },
    confirmationToken: String,
    questions: [
      {
        type: Schema.Types.ObjectId,
        ref: 'question',
        required: 'You must supply a question',
      },
    ],
    votes: {
      like: [
        {
          type: Schema.Types.ObjectId,
          ref: 'question',
        },
      ],
      dislike: [
        {
          type: Schema.Types.ObjectId,
          ref: 'question',
        },
      ],
    },
  },
  {
    toJSON: { virtuals: true },
    toOjbect: { virtuals: true },
  },
);

userSchema.methods.generateJWT = function uSchema1() {
  return jwt.sign(
    {
      _id: this._id,
      username: this.username,
      email: this.email,
      role: this.role,
    },
    process.env.SECRET,
  );
};

userSchema.virtual('gravatar').get(function uSchema2() {
  const hash = md5(this.email);
  return `https://gravatar.com/avatar/${hash}?s=100`;
});

userSchema.virtual('qlists', {
  ref: 'qlist',
  localField: '_id',
  foreignField: 'author',
});

userSchema.virtual('candidates', {
  ref: 'candidate',
  localField: '_id',
  foreignField: 'interviewer',
});

userSchema.plugin(uniqueValidator);

export default mongoose.model('user', userSchema);
