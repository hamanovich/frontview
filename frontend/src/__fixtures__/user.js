export default {
  _id: '5bd2ce32cc8cfa2208379ad5',
  votes: {
    like: [],
    dislike: [],
  },
  notes: 'Please add some notes about yourself',
  role: 'user',
  confirmed: true,
  questions: ['5bd2dedb23c8a03cc87f2fbc'],
  username: 'test',
  email: 'test@test.com',
  passwordDigest: '$2b$10$bLgRlIJSlejDEgJLwsNgV.mlm55s1d7pJKbGH0Hf8boL1Bd7V9Px2',
  confirmationToken: null,
};
