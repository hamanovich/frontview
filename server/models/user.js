import mongoose from 'mongoose';
import md5 from 'md5';

const Schema = mongoose.Schema;

const userSchema = new Schema({
  username: {
    type: String,
    unique: true,
    trim: true,
    required: 'Username field is required'
  },
  email: {
    type: String,
    unique: true,
    lowercase: true,
    trim: true,
    required: 'Email field is required'
  },
  password_digest: {
    type: String,
    required: 'Password field is required'
  },
  first_name: String,
  last_name: String,
  primary_skill: String,
  job_function: String,
  skype: String,
  phone: String,
  notes: {
    type: String,
    default: 'Please add some notes about yourself'
  },
  resetPasswordToken: String,
  resetPasswordExpires: Date
});

userSchema.virtual('gravatar').get(function () {
  const hash = md5(this.email);
  return `https://gravatar.com/avatar/${hash}?s=100`;
});

export default mongoose.model('user', userSchema);