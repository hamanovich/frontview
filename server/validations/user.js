import isEmpty from 'lodash/isEmpty';
import User from '../models/user';

export default async (data, otherValidations) => {
  const { username, email } = data;
  const { errors } = await otherValidations(data);
  const user = await User.find({ $or: [{ username }, { email }] });

  if (user) {
    if (user.username === data.username) {
      errors.username = 'There is a user with such username';
    }

    if (user.email === data.email) {
      errors.email = 'There is a user with such email';
    }
  }

  return {
    errors,
    isValid: isEmpty(errors)
  };
};