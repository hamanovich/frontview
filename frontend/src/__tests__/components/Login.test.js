import React from 'react';
import { shallow } from 'enzyme';

import {
  Login as LoginForm,
  onSubmit,
} from '../../components/LoginPage/Login.tsx';
import user from '../../__fixtures__/user';

describe('<Login/>', () => {
  const initialState = {
    error: '',
    isLoading: false,
  };
  const props = {
    state: initialState,
    addFlashMessage: jest.fn(),
    onSubmit: jest.fn(),
    handleSubmit: fn => fn,
  };
  const loginOK = jest.fn(() => Promise.resolve('Success'));
  const loginFailed = jest.fn(() => Promise.reject(new Error('Failed')));
  const propsHandler = {
    login: loginOK,
    history: {
      push: jest.fn(),
    },
    getUser: jest.fn(),
    setState: jest.fn(),
  };

  const component = shallow(<LoginForm {...props} />);

  it('renders <Login /> component', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders <Login /> component with error state', () => {
    const errorState = { error: 'Test Error', isLoading: false };
    const component = shallow(<LoginForm {...props} state={errorState} />);

    expect(component).toMatchSnapshot();
    expect(component.find('Alert[variant="danger"]').html()).toBe(
      `<div role="alert" class="fade alert alert-danger show">${errorState.error}</div>`,
    );
  });

  describe('when User submits the Login Form and Promise resolves with status 200', () => {
    beforeEach(() => component.find('Form').simulate('submit', user));

    it('invokes onSubmit method', () => {
      expect(props.onSubmit).toHaveBeenCalledWith(user);
    });
  });

  describe('when user tries to submit Login form and status OK', () => {
    it('invokes login props method with `user`', async () => {
      await onSubmit(propsHandler)(user);
      expect(propsHandler.login).toHaveBeenCalledWith(user);
    });

    it('invokes login props method with `user`', async () => {
      await onSubmit({ ...propsHandler, login: loginFailed })(user);
      expect(propsHandler.setState).toHaveBeenCalled();
    });
  });
});
