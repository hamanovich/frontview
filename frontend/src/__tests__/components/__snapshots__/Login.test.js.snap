// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login/> renders <Login /> component 1`] = `
<Fragment>
  <PageHeader
    bsClass="page-header"
  >
    <FontAwesome
      name="user-circle-o"
    />
     Please, login
  </PageHeader>
  <Form
    bsClass="form"
    componentClass="form"
    horizontal={false}
    inline={false}
    noValidate={true}
    onSubmit={[MockFunction]}
  >
    <Field
      component={[Function]}
      htmlFor="identifier"
      label="Username / Email*:"
      name="identifier"
      placeholder="Type your Username or Email"
      type="text"
    />
    <Field
      component={[Function]}
      label="Password*:"
      name="password"
      placeholder="Come up with a password"
      type="password"
    />
    <FormGroup
      bsClass="form-group"
    >
      <FormControlStatic
        bsClass="form-control-static"
        componentClass="p"
      >
        <Link
          replace={false}
          to="/login/forgot"
        >
          Forgot password?
        </Link>
      </FormControlStatic>
    </FormGroup>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsSize="large"
      bsStyle="primary"
      disabled={false}
      type="submit"
    >
      Login
      <FontAwesome
        name="sign-in"
      />
    </Button>
  </Form>
</Fragment>
`;

exports[`<Login/> renders <Login /> component with error state 1`] = `
<Fragment>
  <PageHeader
    bsClass="page-header"
  >
    <FontAwesome
      name="user-circle-o"
    />
     Please, login
  </PageHeader>
  <Form
    bsClass="form"
    componentClass="form"
    horizontal={false}
    inline={false}
    noValidate={true}
    onSubmit={[MockFunction]}
  >
    <Alert
      bsClass="alert"
      bsStyle="danger"
      closeLabel="Close alert"
    >
      Test Error
    </Alert>
    <Field
      component={[Function]}
      htmlFor="identifier"
      label="Username / Email*:"
      name="identifier"
      placeholder="Type your Username or Email"
      type="text"
    />
    <Field
      component={[Function]}
      label="Password*:"
      name="password"
      placeholder="Come up with a password"
      type="password"
    />
    <FormGroup
      bsClass="form-group"
    >
      <FormControlStatic
        bsClass="form-control-static"
        componentClass="p"
      >
        <Link
          replace={false}
          to="/login/forgot"
        >
          Forgot password?
        </Link>
      </FormControlStatic>
    </FormGroup>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsSize="large"
      bsStyle="primary"
      disabled={false}
      type="submit"
    >
      Login
      <FontAwesome
        name="sign-in"
      />
    </Button>
  </Form>
</Fragment>
`;
